tests/corrupted_stamp_test.cpp

```diff
@@ -0,0 +1,318 @@
+        ehdr->timestamp.tnsec = timestamp_distribution(random_generator);
+
+        ehdr->__pad1[0] = 0;
+        ehdr->__pad1[1] = 0;
+        ehdr->__pad1[2] = 0;
+
+        ehdr->__pad2[0] = 0;
+        ehdr->__pad2[1] = 0;
+
+        return buffer;
+    };
+
+    while(number--) {
+        const auto hdr = make_random_header(number);
+        BOOST_REQUIRE_EQUAL(blob_check_corrupted_stamp(hdr.data(), hdr.size()), -EILSEQ);
+    }
+}
+
+data_pointer make_headers_stamp(const headers_test_case &test_case, size_t to_add = 0, size_t to_back_remove = 0) {
+    auto buffer = data_pointer::allocate(HEADERS_SIZE + to_add);
+
+    memset(buffer.data(), 0, HEADERS_SIZE);
+
+    auto dc = buffer.data<eblob_disk_control>();
+    auto ehdr = buffer.skip(sizeof(eblob_disk_control)).data<dnet_ext_list_hdr>();
+
+    dc->flags = test_case.dc_flags;
+    dc->data_size = test_case.dc_data_size;
+    dc->disk_size = test_case.dc_disk_size;
+
+    ehdr->version = test_case.ehdr_version;
+    ehdr->timestamp.tsec = test_case.ehdr_timestamp_seconds;
+    ehdr->timestamp.tnsec = test_case.ehdr_timestamp_seconds ^ 1; // just for completeness.
+
+    memcpy(ehdr->__pad1, test_case.pad1, sizeof(test_case.pad1));
+    memcpy(ehdr->__pad2, test_case.pad2, sizeof(test_case.pad2));
+
+    return data_pointer::copy(buffer.slice(0, HEADERS_SIZE + to_add - to_back_remove));
+}
+
+// Test valid headers with explicitly set fields, most checks for marginal headers fields values.
+void test_explicit_headers() {
+    for(const auto &test_case: marginal_cases) {
+        const auto stamp = make_headers_stamp(test_case);
+
+        BOOST_REQUIRE_EQUAL(blob_check_corrupted_stamp(stamp.data(), stamp.size()), test_case.expected_error_code);
+        BOOST_REQUIRE_EQUAL(blob_check_corrupted_stamp(stamp.data(), stamp.size() - 1), 0);
+    }
+}
+
+// TODO(karapuz): recreate blob file with different offsets.
+void test_read_corrupted_stamp(ioremap::elliptics::newapi::session &s) {
+    namespace ell = ioremap::elliptics;
+    s.set_groups(groups);
+
+    struct test_read_result {
+        ell::async_result<ell::newapi::read_result_entry> async_result;
+        int expected_error_code;
+    };
+
+    std::vector<test_read_result> read_results;
+    read_results.reserve(functional_cases.size());
+
+    const auto make_read_result = [&] (const std::string& key_prefix,
+                                       headers_test_case& test_case,
+                                       int id,
+                                       int ec,
+                                       size_t to_add,
+                                       size_t to_remove)
+     {
+        std::ostringstream key;
+        key << key_prefix << " id: " << id << " expected error: " << ec;
+
+        auto k = key.str();
+        auto data = make_headers_stamp(test_case, to_add, to_remove);
+
+        s.write_data(k, data, 0).wait();
+        auto async = s.read_data(k, 0, 0);
```

Comment:
Let's also check `read`, `read_json` and all three methods with non-zero offset and size.